File: mindcraft_py/connection_handler.py
from __future__ import annotations

import json
import re
from pathlib import Path

_ERROR_DEFINITIONS_PATH = Path(__file__).with_name("connection_errors.json")

with _ERROR_DEFINITIONS_PATH.open("r", encoding="utf-8") as fh:
    ERROR_DEFINITIONS = json.load(fh)


def log(agent_name, msg):
    print(msg)


def parse_kick_reason(reason):
    if not reason:
        return {"type": "unknown", "msg": "Unknown reason (Empty)", "isFatal": True}

    raw = (reason if isinstance(reason, str) else json.dumps(reason)).lower()

    for type_name, definition in ERROR_DEFINITIONS.items():
        if any(keyword in raw for keyword in definition["keywords"]):
            return {
                "type": type_name,
                "msg": definition["msg"],
                "isFatal": definition["isFatal"],
            }

    fallback = raw
    try:
        obj = reason if isinstance(reason, dict) else json.loads(reason)
        fallback = (
            obj.get("translate")
            or obj.get("text")
            or obj.get("value", {}).get("translate")
            or raw
        )
    except Exception:
        pass

    return {"type": "other", "msg": f"Disconnected: {fallback}", "isFatal": True}


def handle_disconnection(agent_name, reason):
    parsed = parse_kick_reason(reason)
    final_msg = f"[LoginGuard] {parsed['msg']}"
    log(agent_name, final_msg)
    return {"type": parsed["type"], "msg": final_msg}


def validate_name_format(name):
    if not name or not re.match(r"^[a-zA-Z0-9_]{3,16}$", name):
        return {
            "success": False,
            "msg": (
                f"[LoginGuard] Invalid name '{name}'. Must be 3-16 "
                "alphanumeric/underscore characters."
            ),
        }
    return {"success": True}

File: src/agent/connection_handler.js
import { sendOutputToServer } from './mindserver_proxy.js';
import { readFileSync } from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ERROR_DEFINITIONS = JSON.parse(
    readFileSync(path.join(__dirname, '..', '..', 'mindcraft_py', 'connection_errors.json'), 'utf8')
);

// Helper to log messages to console (once) and MindServer.
export const log = (agentName, msg) => {
    // Use console.error for visibility in terminal
    console.error(msg);
    try { sendOutputToServer(agentName || 'system', msg); } catch (_) {}
};

// Analyzes the kick reason and returns a full, human-readable sentence.
export function parseKickReason(reason) {
    if (!reason) return { type: 'unknown', msg: 'Unknown reason (Empty)', isFatal: true };
    
    const raw = (typeof reason === 'string' ? reason : JSON.stringify(reason)).toLowerCase();

    // Search for keywords in definitions
    for (const [type, def] of Object.entries(ERROR_DEFINITIONS)) {
        if (def.keywords.some(k => raw.includes(k))) {
            return { type, msg: def.msg, isFatal: def.isFatal };
        }
    }
    
    // Fallback: Extract text from JSON
    let fallback = raw;
    try {
        const obj = typeof reason === 'string' ? JSON.parse(reason) : reason;
        fallback = obj.translate || obj.text || (obj.value?.translate) || raw;
    } catch (_) {}
    
    return { type: 'other', msg: `Disconnected: ${fallback}`, isFatal: true };
}

// Centralized handler for disconnections.
export function handleDisconnection(agentName, reason) {
    const { type, msg } = parseKickReason(reason);
    
    // Format: [LoginGuard] Error Message
    const finalMsg = `[LoginGuard] ${msg}`;
    
    // Only call log once (it handles console printing)
    log(agentName, finalMsg);
    
    return { type, msg: finalMsg };
}

// Validates name format.
export function validateNameFormat(name) {
    if (!name || !/^[a-zA-Z0-9_]{3,16}$/.test(name)) {
        return { 
            success: false, 
            // Added [LoginGuard] prefix here for consistency
            msg: `[LoginGuard] Invalid name '${name}'. Must be 3-16 alphanumeric/underscore characters.` 
        };
    }
    return { success: true };
}
